proposals/0396-observability.md

```diff
@@ -0,0 +1,760 @@
+* Adopting the features for observability should require minimal effort to get started
+* Using advanced features should progressively disclose to more complex systems
+* Observation should be able to handle more than one observed member at once
+* Observation should be able to work with computed properties that reference other properties
+* Observation should be able to work with computed properties that process both get and set to external storage
+* Integration of observation should work in transactions of graphs and not just singular objects
+
+We propose a new standard library module named `Observation` that includes the protocols, types, and macros to implement such a pattern.
+
+Primarily, a type can declare itself as observable simply by using the `@Observable` macro annotation:
+
+```swift
+@Observable public final class MyObject {
+    public var someProperty: String = ""
+    public var someOtherProperty: Int = 0
+    fileprivate var somePrivateProperty: Int = 1
+}
+```
+
+The `@Observable` macro declares and implements conformance to the `Observable` protocol, which includes a set of extension methods to handle observation. The `Observable` protocol supports three ways to interact with changed properties: tracking specific values changes coalesced around iteration, tracking changes coalesced around a specific actor isolation, and interoperating with UI. The first two systems are intended for non-UI uses, while user interface libraries such as SwiftUI can use the last system to allow data changes to flow through to views.
+
+In the simplest case, a client can use the `values(for:)` method to observe changes to a specific property for a given instance.
+
+```swift
+func processChanges(_ object: MyObject) async {
+    for await value in object.values(for: \.someProperty) {
+        print(value)
+    }
+}
+```
+
+This allows users of `Observable` types to observe changes to specific values or an instance as a whole as asynchronous sequences of change events. Because the `values(for:)` method only observes a single property, it provides a sequence of elements of that property's type.
+
+```swift
+object.someProperty = "hello" 
+// prints "hello" in the awaiting loop
+object.someOtherProperty += 1
+// nothing is printed
+```
+
+Observable objects can also provide changes grouped into transactions, which coalesce any changes that are made between suspension points. Transactions are accessed using the `changes(for:)` method, and are delivered isolated to an actor that you provide, or the main actor by default. Because transactions can cover changes to more than one property, the `changes(for:)` method provides a sequence of `TrackedProperties` instances, which can be queried to find the changed properties.
+
+```swift
+func processTransactions(_ object: MyObject) async {
+    for await change in objects.changes(for: [\.someProperty, \.someOtherProperty]) {
+        print(myObject.someProperty, myObject.someOtherProperty)
```

Comment:
Where does `myObject` come from here? Is it possibly a typo and was it meant to be `object` instead?

Reply:
it was.